client/client.go

```diff
@@ -0,0 +1,91 @@
+package client
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrNotFound = errors.New("hardware not found")
+)
+
+// Finder is a type for discovering hardware.
+type Finder interface {
+	ByIP(context.Context, net.IP) (Discoverer, error)
+	ByMAC(context.Context, net.HardwareAddr, net.IP, string) (Discoverer, error)
+}
+
+// WorkflowFinder looks for a Tinkerbell workflow for a given HardwareID.
+type WorkflowFinder interface {
+	HasActiveWorkflow(context.Context, HardwareID) (bool, error)
+}
+
+type Component struct {
+	Type            string      `json:"type"`
+	Name            string      `json:"name"`
+	Vendor          string      `json:"vendor"`
+	Model           string      `json:"model"`
+	Serial          string      `json:"serial"`
+	FirmwareVersion string      `json:"firmware_version"`
+	Data            interface{} `json:"data"`
+}
+
+type ComponentsResponse struct {
+	Components []Component `json:"components"`
+}
+
+// BondingMode is the hardware bonding mode
+type BondingMode int
+
+type HardwareID string
+
+func (hid HardwareID) String() string {
+	return string(hid)
+}
+
+// InstanceState represents the state of an instance (e.g. active)
+type InstanceState string
+
+// Discoverer interface is the base for cacher and tinkerbell hardware discovery
+type Discoverer interface {
+	Instance() *Instance
+	MAC() net.HardwareAddr
+	Mode() string
+	GetIP(addr net.HardwareAddr) IP
+	GetMAC(ip net.IP) net.HardwareAddr
+	DnsServers(mac net.HardwareAddr) []net.IP
+	LeaseTime(mac net.HardwareAddr) time.Duration
+	Hostname() (string, error)
+	Hardware() Hardware
+	SetMAC(mac net.HardwareAddr)
+}
+
+// HardwareState is the hardware state (e.g. provisioning)
+type HardwareState string
+
+// Hardware interface holds primary hardware methods
+type Hardware interface {
```

Comment:
This interface has a bit of a smell about it. I can't quite put my finger on it, but I think its the combination of mac parameterized and non-parmaterized methods. Why does a piece of hardware not know its mac and consequently need a mac for returning things like a kernel path etc?

Reply:
This was moved without changes from `/packet/models.go`. For this PR, I'm not trying to refactor all the business logic, really just focusing on the control flow of how data is fetched from external sources

Reply:
Gotcha, I didn't make the connection.